cupy/core/core.pyx

```diff
@@ -2099,52 +2093,55 @@ cpdef ndarray _adv_getitem(ndarray a, slices):
                 ri = i
 
     if do_transpose:
-        transp = list(range(a.ndim))
-        p = 0
-        for i, s in enumerate(list(slices)):
-            if isinstance(s, ndarray):
-                transp.remove(i)
-                transp.insert(p, i)
-                tmp = slices.pop(i)
-                slices.insert(p, tmp)
-                p += 1
-        a = a.transpose(*transp)
+        transp_a = []
+        transp_b = []
+        slices_a = []
+        slices_b = []
 
+        for i, s in enumerate(slices):
+            if isinstance(s, ndarray):
+                transp_a.append(i)
+                slices_a.append(s)
+            else:
+                transp_b.append(i)
+                slices_b.append(s)
+        a = a.transpose(*(transp_a + transp_b))
+        slices = slices_a + slices_b
         li = 0
-        ri = p - 1
+        ri = len(transp_a) - 1
+
+    a_shepe = a.shape
 
     # flatten the array-indexed dimensions
-    shape = a.shape[:li] +\
-        (internal.prod_ssize_t(a.shape[li:ri+1]),) + a.shape[ri+1:]
-    input_flat = a.reshape(shape)
+    shape = a_shepe[:li] +\
+        (internal.prod_ssize_t(a_shepe[li:ri+1]),) + a_shepe[ri+1:]
+    input_flat = a._reshape(shape)
 
     # build the strides
     strides = [1]
-    for i in range(ri, li, -1):
-        stride = a.shape[i] * strides[0]
-        strides.insert(0, stride)
+    for s in a.shape[ri:li:-1]:
+        strides.insert(0, s * strides[0])
 
     # convert all negative indices to wrap_indices
     for i in range(li, ri+1):
```

Comment:
`range` is special function in Cython.
Please use `range`  in pyx file.
http://cython.readthedocs.io/en/latest/src/reference/language_basics.html#for-loops